server/bleep/src/webserver/answer.rs

```diff
@@ -309,151 +399,250 @@ async fn _handle(
         None
     };
 
+    let semantic = app
+        .semantic
+        .clone()
+        .ok_or_else(|| Error::new(ErrorKind::Configuration, "Qdrant not configured"))?;
+
     let answer_api_client = semantic.build_answer_api_client(
         state,
         format!("{}/q", app.config.answer_api_url).as_str(),
-        query,
         5,
         answer_bearer.clone(),
     );
 
-    let select_prompt = answer_api_client.build_select_prompt(&snippets);
+    let mut progress = app
+        .with_prior_conversation(&thread_id, |history| {
+            if history.is_empty() {
+                None
+            } else {
+                Some(query.clone())
+            }
+        })
+        .map(AnswerProgress::Rephrase)
+        .unwrap_or(AnswerProgress::Rephrase(query.clone()));
 
-    analytics_event
+    event
+        .write()
+        .await
         .stages
-        .push(Stage::new("select prompt", &select_prompt).with_time(stop_watch.lap()));
-
-    let relevant_snippet_index = answer_api_client
-        .select_snippet(&select_prompt)
-        .await?
-        .trim()
-        .to_string()
-        .clone();
-
-    info!("Relevant snippet index: {}", &relevant_snippet_index);
-
-    let mut relevant_snippet_index = relevant_snippet_index
-        .parse::<usize>()
-        .map_err(Error::internal)?;
-
-    analytics_event.stages.push(
-        Stage::new("relevant snippet index", relevant_snippet_index).with_time(stop_watch.lap()),
-    );
-
-    let (answer_path, stream) = if relevant_snippet_index > 0 {
-        relevant_snippet_index -= 1; // return to 0-indexing
-        let relevant_snippet = snippets
-            .get(relevant_snippet_index)
-            .ok_or_else(|| Error::internal("answer-api returned out-of-bounds index"))?;
-
-        // grow the snippet by 60 lines above and below, we have sufficient space
-        // to grow this snippet by 10 times its original size (15 to 150)
-
-        let repo_ref = &relevant_snippet
-            .repo_ref
-            .parse::<RepoRef>()
-            .map_err(Error::internal)?;
-        let doc = app
-            .indexes
-            .file
-            .by_path(repo_ref, &relevant_snippet.relative_path)
-            .await
-            .map_err(Error::internal)?;
-
-        let mut grow_size = 40;
-        let grown_text = loop {
-            if let Some(grown_text) = grow(&doc, relevant_snippet, grow_size) {
-                let token_count = semantic.gpt2_token_count(&grown_text);
-                info!(%grow_size, %token_count, "growing ...");
-                if token_count > 2000 || grow_size > 100 {
-                    break grown_text;
+        .push(Stage::new("start", &query).with_time(stop_watch.lap()));
+
+    loop {
```

Comment:
This is an unbounded `loop` inside a very long function, which make this logic inherently hard to reason about. It would be good to either make this shorter, or change this to `loop` to something with explicit bounds.